Assets/MixedRealityToolkit.Examples/Experimental/Audio/Scripts/MicrophoneAmplitudeDemo.cs

```diff
@@ -0,0 +1,208 @@
+        }
+
+        private void OnDestroy()
+        {
+            // Stop the microphone stream.
+            WindowsMicrophoneStreamErrorCode result = micStream.StopStream();
+            if (result != WindowsMicrophoneStreamErrorCode.Success)
+            {
+                Debug.Log($"Failed to stop the microphone stream. {result}");
+            }
+
+            // Uninitialize the microphone stream.
+            micStream.Uninitialize();
+            micStream = null;
+
+#if UNITY_EDITOR
+            // Restore the initial material settings.
+            if (VisibleMaterial != null)
+            {
+                VisibleMaterial.SetColor("_WireColor", defaultMaterialColor);
+                VisibleMaterial.SetInt("_WireThickness", defaultWireThickness);
+            }
+#endif // UNITY_EDITOR
+        }
+
+        private void OnDisable()
+        {
+            // Pause the microphone stream.
+            WindowsMicrophoneStreamErrorCode result = micStream.Pause();
+            if (result != WindowsMicrophoneStreamErrorCode.Success)
+            {
+                Debug.Log($"Failed to pause the microphone stream. {result}");
+            }
+        }
+
+        private void OnEnable()
+        {
+            // Resume the microphone stream.
+            WindowsMicrophoneStreamErrorCode result = micStream.Resume();
+            if (result != WindowsMicrophoneStreamErrorCode.Success)
+            {
+                Debug.Log($"Failed to resume the microphone stream. {result}");
+            }
+        }
+
+        private static int maxWireThickness = 750;
+
+        private void Update()
+        {
+            if (VisibleMaterial != null)
+            {
+                // Artificially increase the amplitude to make the visible effect more pronounced.
+                int wireThickness = (int)(averageAmplitude * amplitudeBoostFactor * maxWireThickness);
+                wireThickness = Mathf.Clamp(wireThickness, 0, maxWireThickness);
+                VisibleMaterial.SetInt("_WireThickness", wireThickness);
+            }
+        }
+
+        private void OnAudioFilterRead(float[] buffer, int numChannels)
+        {
+            // Read the microphone stream data.
+            WindowsMicrophoneStreamErrorCode result = micStream.ReadAudioFrame(buffer, numChannels);
+            if (result != WindowsMicrophoneStreamErrorCode.Success)
+            {
+                Debug.Log($"Failed to read the microphone stream data. {result}");
+            }
+
+            float sumOfValues = 0;
+
+            // Calculate this frame's average amplitude.
+            for (int i = 0; i < buffer.Length; i++)
+            {
+                if (float.IsNaN(buffer[i]))
+                {
+                    buffer[i] = 0;
+                }
+
+                buffer[i] = Mathf.Clamp(buffer[i], -1.0f, 1.0f);
+                sumOfValues += Mathf.Clamp01(Mathf.Abs(buffer[i]));
+            }
+
+            averageAmplitude = sumOfValues / buffer.Length;
```

Comment:
Is it possible for the buffer to have zero length?

Reply:
If so we should probably enclose this entire thing in a zero check

---
In reply to: [308807929](https://github.com/microsoft/MixedRealityToolkit-Unity/pull/5455#discussion_r308807929) [](ancestors = 308807929)